File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/AssessmentSystem.java
import java.util.*;
import java.util.Optional;
import java.util.Map;

class Student {
    private Map<String, Integer> courses;
    private int grade;
    private String major;
    private List<Double> scores;

    public Student(int grade, String major) {
        this.grade = grade;
        this.major = major;
        this.courses = new HashMap<>();
        this.scores = new ArrayList<>();
    }

    public Map<String, Integer> getCourses() {
        return courses;
    }
}

public class AssessmentSystem {
    private Map<String, Student> students = new HashMap<>();

    public void addStudent(String name, int grade, String major) {
        students.put(name, new Student(grade, major));
    }

    public void addCourseScore(String name, String course, int score) {
        if (students.containsKey(name)) {
            students.get(name).getCourses().put(course, score);
        }
    }

    public Optional<Double> getGPA(String name) {
        if (students.containsKey(name)) {
            Map<String, Integer> courses = students.get(name).getCourses();
            if (!courses.isEmpty()) {
                double totalScore = courses.values().stream()
                    .mapToDouble(Integer::doubleValue)
                    .sum();
                return Optional.of(totalScore / courses.size());
            }
        }
        return Optional.empty();
    }

    public List<String> getAllStudentsWithFailCourse() {
        List<String> studentsWithFail = new ArrayList<>();
        for (Map.Entry<String, Student> entry : students.entrySet()) {
            String name = entry.getKey();
            Student student = entry.getValue();
            boolean hasFailing = student.getCourses().entrySet().stream()
                .anyMatch(course -> course.getValue() < 60);
            if (hasFailing) {
                studentsWithFail.add(name);
            }
        }
        return studentsWithFail;
    }

    public Optional<Double> getCourseAverage(String course) {
        double total = 0;
        int count = 0;
        for (Student student : students.values()) {
            Integer grade = student.getCourses().get(course);
            if (grade != null) {
                total += grade;
                count++;
            }
        }
        return count > 0 ? Optional.of(total / count) : Optional.empty();
    }

    public Optional<String> getTopStudent() {
        Optional<String> topStudent = Optional.empty();
        Optional<Double> highestGpa = Optional.empty();
        for (Map.Entry<String, Student> entry : students.entrySet()) {
            String name = entry.getKey();
            Optional<Double> gpa = getGPA(name);
            if (gpa.isPresent() &&
                (!highestGpa.isPresent() || gpa.get() > highestGpa.get())) {
                highestGpa = gpa;
                topStudent = Optional.of(name);
            }
        }
        return topStudent;
    }

    public Map<String, Student> getStudents() {
        return students;
    }
}

File: ClassEval_T/cpp/solution/AssessmentSystem.cpp

#include <numeric>
#include <algorithm>

#include <string>
#include <map>
#include <vector>
#include <optional>


struct Student {
    int grade;
    std::string major;
    std::map<std::string, int> courses;
};


class AssessmentSystem {
public:
    void add_student(const std::string& name, int grade, const std::string& major);
    void add_course_score(const std::string& name, const std::string& course, int score);
    std::optional<double> get_gpa(const std::string& name) const;
    std::vector<std::string> get_all_students_with_fail_course() const;
    std::optional<double> get_course_average(const std::string& course) const;
    std::optional<std::string> get_top_student() const;
    std::map<std::string, Student> students;
};

void AssessmentSystem::add_student(const std::string& name, int grade, const std::string& major) {
    students[name] = { grade, major, {} };
}

void AssessmentSystem::add_course_score(const std::string& name, const std::string& course, int score) {
    auto it = students.find(name);
    if (it != students.end()) {
        it->second.courses[course] = score;
    }
}

std::optional<double> AssessmentSystem::get_gpa(const std::string& name) const {
    auto it = students.find(name);
    if (it != students.end()) {
        const auto& courses = it->second.courses;
        if (!courses.empty()) {
            double total_score = std::accumulate(courses.begin(), courses.end(), 0.0,
                [](double sum, const auto& course) { return sum + course.second; });
            return total_score / courses.size();
        }
    }
    return std::nullopt;
}

std::vector<std::string> AssessmentSystem::get_all_students_with_fail_course() const {
    std::vector<std::string> students_with_fail;
    for (const auto& [name, student] : students) {
        if (std::any_of(student.courses.begin(), student.courses.end(), [](const auto& course) { return course.second < 60; })) {
            students_with_fail.push_back(name);
        }
    }
    return students_with_fail;
}

std::optional<double> AssessmentSystem::get_course_average(const std::string& course) const {
    double total = 0;
    int count = 0;
    for (const auto& [name, student] : students) {
        auto it = student.courses.find(course);
        if (it != student.courses.end()) {
            total += it->second;
            ++count;
        }
    }
    return (count > 0) ? std::optional<double>(total / count) : std::nullopt;
}

std::optional<std::string> AssessmentSystem::get_top_student() const {
    std::optional<std::string> top_student;
    std::optional<double> highest_gpa;
    for (const auto& [name, student] : students) {
        auto gpa = get_gpa(name);
        if (gpa && (!highest_gpa || *gpa > *highest_gpa)) {
            highest_gpa = gpa;
            top_student = name;
        }
    }
    return top_student;
}
